spec/libraries/duplicator_spec.rb

```diff
@@ -0,0 +1,763 @@
+          dup_c5 = dup_c3.children[0]
+
+          expect(dup_c1.children.length).to eq(1)
+          expect(dup_c2.children.length).to eq(1)
+          expect(dup_c3.children.length).to eq(1)
+          expect(dup_c5.children.length).to eq(0)
+        end
+
+        it 'duplicates sub-graph from c3' do
+          duplicator = Duplicator.new
+          dup_c3 = duplicator.duplicate(@c3)
+
+          dup_c4 = dup_c3.children[0].children[0]
+
+          expect(dup_c3).to_not be(@c3)
+          expect(dup_c4.children[0]).to eq(@s1)
+          # check cycle
+          expect(dup_c4.children[1]).to be(dup_c3)
+        end
+      end
+
+      context 'when an array of objects is duplicated' do
+        before :each do
+          create_cyclic_graph
+          create_second_cyclic_graph
+          @duplicator = Duplicator.new
+        end
+
+        it 'duplicates objects mentioned twice without creating extras' do
+          duplicated_stuff = @duplicator.duplicate([@c1, @c3])
+
+          dup_c3 = duplicated_stuff[0].children[0].children[0]
+
+          expect(duplicated_stuff.length).to eq(2)
+          expect(duplicated_stuff[0]).to eq(@c1)
+          expect(duplicated_stuff[0]).to_not be(@c1)
+          expect(duplicated_stuff[1]).to eq(@c3)
+          expect(duplicated_stuff[1]).to_not be(@c3)
+          expect(duplicated_stuff[1]).to be(dup_c3)
+        end
+
+        it 'duplicates disjoint graphs' do
+          duplicated_stuff = @duplicator.duplicate([@c1, @c21])
+
+          expect(duplicated_stuff.length).to eq(2)
+          expect(duplicated_stuff[0]).to eq(@c1)
+          expect(duplicated_stuff[0]).to_not be(@c1)
+          expect(duplicated_stuff[1]).to eq(@c21)
+          expect(duplicated_stuff[1]).to_not be(@c21)
+        end
+      end
+
+      context 'when joined graphs are duplicated' do
+        before :each do
+          create_cyclic_graph
+          create_second_cyclic_graph
+          # join graphs
+          c1_children = @c1.children
+          c1_children << @c21
+          @c1.instance_variable_set(:@children, c1_children)
+
+          @duplicator = Duplicator.new
+        end
+
+        it 'duplicates all objects' do
+          @duplicator.duplicate(@c1)
+
+          duplicated_objects = @duplicator.instance_variable_get(:@duplicated_objects)
+          expect(duplicated_objects.length).to eq(9)
+        end
+
+        it 'duplicates cyclically joined graphs' do
+          # join from c21 to c1
+          c21_children = @c21.children
+          c21_children << @c1
+          @c21.instance_variable_set(:@children, c21_children)
+
+          @duplicator.duplicate(@c21)
+
+          duplicated_objects = @duplicator.instance_variable_get(:@duplicated_objects)
+          expect(duplicated_objects.length).to eq(9)
+        end
+      end
+    end
+  end
+
+  context 'when ActiveRecord objects' do
+    class SimpleActiveRecord < ActiveRecord::Base
+      def initialize_duplicate(_duplicator, other)
```

Comment:
Unused method argument - `other`. If it's necessary, use `_` or `_other` as an argument name to indicate that it won't be used. You can also write as `initialize_duplicate(*)` if you want the method to accept any arguments but don't care about them.